src/OpenTelemetry.Exporter.Geneva/Internal/MsgPack/MsgPackTraceExporter.cs

```diff
@@ -320,71 +299,154 @@ internal static bool CacheIfPartOfHttpUrl(KeyValuePair<string, object?> entry, o
+        this.timestampPatchIndex = cursor;
+        cursor += 12; // reserve 12 bytes for the timestamp
+
+        cursor = MessagePackSerializer.WriteMapHeader(buffer, cursor, ushort.MaxValue); // Note: always use Map16 for perf consideration
+        this.mapSizePatchIndex = cursor - 2;
+
+        this.prepopulatedFieldsCount = 0;
+
+        // TODO: Do we support PartB as well?
+        // Part A - core envelope
+        cursor = AddPartAField(buffer, cursor, Schema.V40.PartA.Name, this.partAName);
+        this.prepopulatedFieldsCount++;
+
+        cursor = AddPartAField(buffer, cursor, Schema.V40.PartA.Ver, "4.0");
+        this.prepopulatedFieldsCount++;
+
+        var resourceAttributes = this.resourceProvider().Attributes;
+
+        if (this.resourceFieldNames != null)
+        {
+            // if ResourceFieldNames is set, it overrides the existing prepopulated fields setting.
+            this.prepopulatedFields = [];
+        }
+
+        // this is guaranteed to not be null because it's set in the constructor
+        Guard.ThrowIfNull(this.prepopulatedFields);
+
+        foreach (var resourceAttribute in resourceAttributes)
         {
-            string key = entry.Key;
-            bool isDedicatedField = false;
+            var key = resourceAttribute.Key;
+            var value = resourceAttribute.Value;
 
-            if (entry.Value is string resourceValue)
+            var isWantedAttribute = false;
+            if (this.resourceFieldNames != null)
             {
-                switch (key)
+                // this might seem inefficient, but it's only run once and I don't expect there to be many resource attributes
+                foreach (var wantedAttribute in this.resourceFieldNames!)
                 {
-                    case "service.name":
-                        key = Schema.V40.PartA.Extensions.Cloud.Role;
-                        isDedicatedField = true;
-                        break;
-                    case "service.instanceId":
-                        key = Schema.V40.PartA.Extensions.Cloud.RoleInstance;
-                        isDedicatedField = true;
+                    if (wantedAttribute == key)
+                    {
+                        switch (value)
+                        {
+                            case bool:
+                            case byte:
+                            case sbyte:
+                            case short:
+                            case ushort:
+                            case int:
+                            case uint:
+                            case long:
+                            case ulong:
+                            case float:
+                            case double:
+                            case string:
+                                break;
+                            case null:
+                                throw new ArgumentNullException(key, "Resource attribute must not have a null value.");
```

Comment:
This could crash customer's app.

Reply:
We need to handle this, and follow the principle of never throw exception. (except in the ctor() )

Reply:
Log to event source - https://github.com/mattsains/opentelemetry-dotnet-contrib/blob/main/src/OpenTelemetry.Exporter.Geneva/Internal/ExporterEventSource.cs

Reply:
Maybe put `"<NULL>"` as the value, so at least users know something is broken... 

I need to recheck to see if this is actually going to crash user app or just fail the export? Even failing entire export is bad just because of one attribute, so above is probably better..

Reply:
> Log to event source - https://github.com/mattsains/opentelemetry-dotnet-contrib/blob/main/src/OpenTelemetry.Exporter.Geneva/Internal/ExporterEventSource.cs

I wouldn't suggest that..The exporter is designed to be re-entrant safe... so logging here could cause stackoverflow, if someone enables a EventSource->ILogger bridge.. Low possibility, but as a general principle, we should not log here. If we really need to do logging, then it must be under SuppressionGuard.


Rust has a comment on this: https://github.com/open-telemetry/opentelemetry-rust-contrib/blob/main/opentelemetry-user-events-logs/src/logs/processor.rs#L50-L55

Reply:
Raj has discovered that it's not possible to have null in a resource value: https://github.com/open-telemetry/opentelemetry-dotnet/blob/49bfa23e825cf7320cf1c70f671313600a58d903/test/OpenTelemetry.Tests/Resources/ResourceTests.cs#L34